src/Kernels/MetropolisHastingsLazy.hs

```diff
@@ -15,34 +15,56 @@ propose x = state $ do
   where
     stepsize = 0.9
 
+propose' :: State (StdGen, Double) Double
+propose' = do
+  (gen, current) <- get
+  let (proposal, newGen) = uniformR (current - stepsize, current + stepsize) gen
+  put(newGen, current)
+  return proposal
+  where
+    stepsize = 0.9
+
 acceptOrReject :: Double -> Double -> State StdGen Double
 acceptOrReject current proposal = state $ do
   (rand, newGen) <- uniformR (0.0, 1.0)
   if rand < acceptanceProbability current proposal
    then return (proposal, newGen)
    else return (current, newGen)
 
+acceptOrReject' :: Double -> State (StdGen, Double) Double
+acceptOrReject' proposal = do
+  (gen, current) <- get
+  let (rand, newGen) = uniformR (0.0, 1.0) gen
+  let nextValue = if rand < acceptanceProbability current proposal
+                   then proposal
+                   else current
+  put (newGen, nextValue)
+  fmap snd get
```

Comment:
Is there a reason for not using `return nextValue` here?

Reply:
@tomatitito no reason. `return nextValue` is better.

Reply:
@froth Ok cool, that's what threw me off! Also, I think we should `return proposal` while still putting `(newGen, nextValue)`. It's important that `nextValue` is in the state, so it can be retrieved when building up the markov chain. However, `proposal` is the value that is computed, even though it might not be used. I think it doesn't matter for the result (haven't tried yet, though), but it better reflects the algorithm.

Reply:
I think the result is ignored by replicateM, yes. That would mean only the last value returned would change. But I think we are only interested in the chain anyways, right?

Reply:
I think we could also do return () without changing much?

Reply:
Yes, we are only interested in the chain. So you are right, we could just discard the computed value altogether and return (). But I think that's what `put` does for us: setting the state and returning `()`.

Reply:
ah nice, learned something new today :)

Reply:
@tomatitito  It does not work as expected. If I return () (or put) the signature of transition' becomes `transition' :: State (StdGen, Double) ()` and thereby the signature sampleStateT' becomes `Double -> StdGen -> Int -> ([()], (StdGen, Double))` and by that we loose our chain. Therefore I also think that it is indeed important to `return nextValue`